arangod/Aql/SortExecutor.cpp

```diff
@@ -45,18 +45,39 @@ class OurLessThan {
       : _vpackOptions(options), _input(input), _sortRegisters(sortRegisters) {}
 
   bool operator()(AqlItemMatrix::RowIndex const& a, AqlItemMatrix::RowIndex const& b) const {
-    InputAqlItemRow left = _input.getRow(a);
-    InputAqlItemRow right = _input.getRow(b);
+    InputAqlItemRow const& left = _input.getRow(a);
+    InputAqlItemRow const& right = _input.getRow(b);
     for (auto const& reg : _sortRegisters) {
       AqlValue const& lhs = left.getValue(reg.reg);
       AqlValue const& rhs = right.getValue(reg.reg);
 
       int const cmp = AqlValue::Compare(_vpackOptions, lhs, rhs, true);
 
-      if (cmp < 0) {
-        return reg.asc;
-      } else if (cmp > 0) {
-        return !reg.asc;
+      if (cmp != 0) {
+        /*
+          the expression
+
+            return ((cmp < 0) ^ (!_sortRegister.asc));
+          
+          should produce the same result as its more complex equivalent
+
+            if (cmp < 0) {
+              return reg.asc;
+            } else if (cmp > 0) {
+              return !reg.asc;
+            }
+
+          but it is translated into assembly with one branch instead of two.
```

Comment:
Have you tried that? Because https://godbolt.org/z/vxdASG looks differently. Are there any performance tests that compare the two code pieces?

Reply:
Yes, I tried it with compiler explorer

Reply:
The code you put there is not 100% equivalent.
It is not
```
    if (cmp < 0)  {
        return asc;
    } else {
        return !asc;
    }
```
but it is
```
    if (cmp < 0)  {
        return asc;
    } else if (cmp > 0) {
        return !asc;
    }
    return false;
}
```

Reply:
No, it is equivalent under the assumption that `cmp != 0`. Which your new code does as well.

Reply:
Yes, but `cmp` can be 0 as well.
Try this one, which is more realistic: https://godbolt.org/z/2weN94

Reply:
works as well :D https://godbolt.org/z/joSsZ9 stupid compilers. Have you done performance tests? I would be interested in the impact.

Reply:
I find it astonishing that the compiler does not use a conditional move instruction..!

Godbolt is nice, but in most cases (like this one) you have to use a simplified piece of code. So in these cases I usually also use [IDA Pro](https://www.hex-rays.com/products/ida/support/download_freeware/) to take a look at the actual code.

Reply:
I was expecting that, too. And that was what I was in my first test with simplified code. Maybe this is already too complex. `clang` handles it better, I think. (at least it uses `sete` and stuff)